tensorflow/compiler/jit/kernels/xla_ops.cc

```diff
@@ -520,11 +538,25 @@ void XlaRunOp::Compute(OpKernelContext* ctx) {
         tensorflow::profiler::TraceMeLevel::kInfo);
 
     for (auto& p : closure.resource_var_snapshots()) {
-      snapshot_ptrs.emplace(p.first,
-                            p.second.has_value() ? &p.second.value() : nullptr);
+      const Tensor* tensor_ptr = nullptr;
+      if (p.second.has_value()) {
+        if (resource_var_ptrs.count(p.first) &&
+            resource_var_ptrs[p.first]->SharesBufferWith(p.second.value())) {
```

Comment:
Could you expand why the logic here is different from the one in XlaLocalLaunchOp? Maybe we could unify these two?

Reply:
The difference is due to an additional issue in auto-clustering that the compilation and run are in two ops (`XlaRunOp` and `XlaCompileOp`). So, a snapshot of tensors is stored by `XlaCompileOp` to maintain the shape consistency. The snapshotted tensors need to be used as inputs if the shapes of the `Vars` change. In contrast, snapshot is not needed for manual clustering.

Here, we check if some snapshotted tensors actually are the copies of the tensors in `Vars`. If yes, we release the snapshotted tensors (i.e., copies of the Var tensors) and in effect the Var tensors' ref counts will decrement by one. This will expose the opportunities for `tensor_donation`. In contrast, if we don't release the snapshotted tensors, `tensor_donation` will never happen because the ref counts will at least be 2 (i.e., the copy due to snapshot _and_ the tensor stored in the Var).

Note that the snapshotted tensors will not be copies of the Var tensors if the Vars are written.

Overall, `XlaCompileOp`/`XlaRunOp` need to deal with the snapshotted tensors for shape consistency. If the shape actually do not change and the underlying tensors are the same, we want the buffer donation to happen.